### PR TITLE
abseil: force using non-std types

### PR DESCRIPTION
**Pull request by mattklein123**
Force using absl optional/string_view in most places. Allow use of std::string_view where
required for extensions, in particular WASM.

Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@PiotrSikora this fails with:
```
external/proxy_wasm_cpp_host/include/proxy-wasm/context.h:219:8: note: hidden overloaded virtual function 'proxy_wasm::ContextBase::error' declared here: type mismatch at 1st parameter ('std::string_view' (aka 'basic_string_view<char>') vs 'absl::string_view')
  void error(std::string_view message) override {
```

Is there any chance the WASM cpp code could use absl::string_view so we can land this change?

**Comment by PiotrSikora** [comment]
> @PiotrSikora this fails with:
> 
> ```
> external/proxy_wasm_cpp_host/include/proxy-wasm/context.h:219:8: note: hidden overloaded virtual function 'proxy_wasm::ContextBase::error' declared here: type mismatch at 1st parameter ('std::string_view' (aka 'basic_string_view<char>') vs 'absl::string_view')
>   void error(std::string_view message) override {
> ```
> 
> Is there any chance the WASM cpp code could use absl::string_view so we can land this change?

We used to have support for abseil as an optional dependency, but it was removed after Envoy moved to C++17, since there was no need for it anymore (see: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/41 and https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/pull/38).

I don't mind re-adding abseil to Proxy-Wasm C++ Host if Envoy needs it, but I'm afraid that it's going to turn into a need for abseil in Proxy-Wasm C++ SDK, which definitely shouldn't have a hard dependency on it, so we need to add it as an optional dependency, and default to `std` types.

cc @mathetake

**Comment by mattklein123** [comment]
> I don't mind re-adding abseil to Proxy-Wasm C++ Host if Envoy needs it, but I'm afraid that it's going to turn into a need for abseil in Proxy-Wasm C++ SDK, which definitely shouldn't have a hard dependency on it, so we need to add it as an optional dependency, and default to std types.

Sure this sounds reasonable to me. @envoyproxy/maintainers what do we think about this change overall? IMO it's worth it to pin to the internal types and not using the std types, especially given the recent security issue with string_view.

**Comment by ggreenway** [comment]
I'm in favor of this change. Having different implementations on different platforms, with different behavior (eg nullptr allowed vs not allowed), is not good.

**Comment by jmarantz** [comment]
@PiotrSikora with C++-17 I would think we'd still want absl as a dependency for some of the functionality it adds beyond what's being pulled into std::.

Examples off the top of my head
 - better mutexes/condvars/await including contention detection
 - better time abstraction (though admittedly Envoy doesn't use that much)
 - string joining, splitting, replace, concat, etc
 - spans

I think not all of that is destined to ever go into std::.

**Comment by PiotrSikora** [comment]
> I'm in favor of this change. Having different implementations on different platforms, with different behavior (eg nullptr allowed vs not allowed), is not good.

Note that abseil is the different one here, the rest of the C++ world doesn't accept `nullptr` as `std::string_view`.

AFAICT, MSVC doesn't allow it either (see: [string_view constructor length check nullptr](https://developercommunity.visualstudio.com/t/string-view-constructor-length-check-nullptr/1023937)), so I'm a bit puzzled by your comment regarding different behavior on different platforms.

FWIW, I agree that allowing `nullptr` is a safer behavior, but that's not what C++17 defines.

> @PiotrSikora with C++-17 I would think we'd still want absl as a dependency for some of the functionality it adds beyond what's being pulled into std::.
> 
> Examples off the top of my head
> 
> * better mutexes/condvars/await including contention detection
> * better time abstraction (though admittedly Envoy doesn't use that much)
> * string joining, splitting, replace, concat, etc
> * spans
> * variants
> * optionals
> 
> I think not all of that is destined to ever go into std::.

In Proxy-Wasm C++ Host code (which my comment was referring to), we were only using `absl::string_view` and `absl::optional`, both of which are included in C++17, so the extra dependency didn't provide any value.

I'd love to use `absl::span<uint8_t>` instead of `absl::string_view` for non-string data, but Envoy doesn't use that, so it would be a bunch of unnecessary conversions in the Host code.

But like I said, I'm fine with accepting abseil as an optional dependency if Envoy switches back to non-std types, but that's the only justification for having it there at this point.

**Comment by ggreenway** [comment]
> your comment regarding different behavior on different platforms

Because on some platforms the abseil implementation was being used, and on some the c++17 std library was being used. IIRC, envoy-mobile needed the abseil implementation somewhere. 

I am not aware of any c++ std library implementation that accepts a nullptr in this case without crashing.

**Comment by lizan** [comment]
> I am not aware of any c++ std library implementation that accepts a nullptr in this case without crashing.

libstdc++ does allow nullptr without crashing.

**Comment by antoniovicente** [comment]
> > I am not aware of any c++ std library implementation that accepts a nullptr in this case without crashing.
> 
> libstdc++ does allow nullptr without crashing.

Can you expand on this?  I was lead to believe strlen(nullptr) results in a crash.  Does the libstdc++ string_view avoid the call to strlen or have a strlen that checks for nullptr?

**Comment by mattklein123** [comment]
It sounds like we don't have consensus here on whether this change is wanted or not. Do we want to adhere more to the C++ standard? Or use the built-in safety/consistency of the abseil version? If we stick with the C++ standard is there some way clang-tidy can catch a C string being passed into a string_view without wrapping?

**Comment by ggreenway** [comment]
> It sounds like we don't have consensus here on whether this change is wanted or not. Do we want to adhere more to the C++ standard? Or use the built-in safety/consistency of the abseil version? If we stick with the C++ standard is there some way clang-tidy can catch a C string being passed into a string_view without wrapping?

I prefer the safer version. I'm not concerned about the performance cost of an extra nullptr check.

I don't think static analysis will in practice help much, because it doesn't know the difference between a possibly null `char*` and a guaranteed-not-null `char*`.

**Comment by antoniovicente** [comment]
I think that using the safer version for now is a good idea.  Bonus points for reaching out to abseil about the possibility of using the nullptr safe version in the StrCat implementation.

**Review by antoniovicente — APPROVED** [review]

**Comment by mattklein123** [comment]
OK sounds like we are roughly agreed that we should proceed with this change. @PiotrSikora are you willing to bring back abseil support in the WASM code (as a compile time option)?

**Comment by mattklein123** [comment]
OK it sounds like we have some consensus here on moving forward. @PiotrSikora are you willing to bring back the option in proxy-wasm to use abseil and then I can finish this change? Or would you prefer I do that change?

**Comment by PiotrSikora** [comment]
@mattklein123 if you could do it, that would be great. Thanks!

**Comment by mattklein123** [comment]
I'm coming back to this and now wasm_cpp is not the only problem, the skywalking SDK also has the same problem. I'm concerned that we may be playing whackamole here given that we support C++17. We would essentially have to force all deps to use abseil. Now that I think about it I'm not sure this is reasonable. @lizan @antoniovicente @ggreenway thoughts?

**Comment by PiotrSikora** [comment]
> I'm coming back to this and now wasm_cpp is not the only problem, the skywalking SDK also has the same problem. I'm concerned that we may be playing whackamole here given that we support C++17. We would essentially have to force all deps to use abseil. Now that I think about it I'm not sure this is reasonable. @lizan @antoniovicente @ggreenway thoughts?

I think that simply allowing use of `std::string_view`, so that Envoy core uses `absl::string_view`, and extensions that inherit from classes in external C++ libraries (like Wasm and Skywalking) use `std::string_view`, should also work?

**Comment by mattklein123** [comment]
> I think that simply allowing use of std::string_view, so that Envoy core uses absl::string_view, and extensions that inherit from classes in external C++ libraries (like Wasm and Skywalking) use std::string_view, should also work?

Yes I agree, that would be a fallback position and if we want to go that route that would be fine. I think we would have to relax or linting or NOLINT all sites that use it in the core code, but I can look into that.

**Comment by ggreenway** [comment]
> I'm coming back to this and now wasm_cpp is not the only problem, the skywalking SDK also has the same problem. I'm concerned that we may be playing whackamole here given that we support C++17. We would essentially have to force all deps to use abseil. Now that I think about it I'm not sure this is reasonable. @lizan @antoniovicente @ggreenway thoughts?

Right now, envoy uses absl all over the place, including in interfaces. Unless we want to stop doing that, I think extensions will require absl.

I believe we still require absl types because not all supported platforms have full c++17 support (envoy mobile? can't remember the details).

But if leaving absl in "use the c++17 version if available" mode makes it easier on everyone, I'm fine with that as a compromise.

**Comment by mattklein123** [comment]
> But if leaving absl in "use the c++17 version if available" mode makes it easier on everyone, I'm fine with that as a compromise.

I'm working on this now and I think this is the way it has to be as the problem reaches everywhere, including protobuf. I have a patch I'm relatively happy with. I will share it by EOD.

**Comment by mattklein123** [comment]
Sorry, I meant I'm leaving it in force mode and then doing conversions where required. I think this makes things more clear anyway but you all can decide.

**Comment by mattklein123** [comment]
@PiotrSikora @ggreenway @lizan @antoniovicente PTAL and let me know what you think. I'm happy to drop this if we think it's not worthwhile, but I think this is relatively clean and gets the job done for the most part.

File: tools/code_format/check_format.py
Comment (PiotrSikora) [inline tools/code_format/check_format.py:126]: Nit: add comment describing this allowlist.

**Review by PiotrSikora — DISMISSED** [review]
Thanks, I think this makes the most sense!

**Review by ggreenway — CHANGES_REQUESTED** [review]
LGTM, modulo @PiotrSikora 's request for a comment in the linter

**Comment by ggreenway** [comment]
Merge conflict.

/wait

**Comment by mattklein123** [comment]
@ggreenway @PiotrSikora updated PTAL

**Review by ggreenway — APPROVED** [review]

**Pull request merged**
